### DISCUSSION TITLE
Swipe on trackpad to switch tabs

### DISCUSSION THREAD

**Discussion by Nerixyz** [Chatterino/chatterino2#4006] [Ideas]
Most browsers have trackpad gestures to go back and forth in history (sideways scroll, see video). In chatterino this could be used to quickly switch tabs.

[firefox.webm](https://user-images.githubusercontent.com/19953266/192090328-e0900f04-26fb-43ef-ad18-99f621f34822.webm)

**Comment by pajlada** [comment]
Could you check if it's possible to catch this as a key event?
I'd really like this in the custom hotkey functionality rather than something new

**Reply by Nerixyz** [reply]
There's no key event on a swipe. [`QSwipeGesture`](https://doc.qt.io/qt-5/qswipegesture.html) seems to provide information about swiping. It's dispatched through a [`QGestureEvent`](https://doc.qt.io/qt-5/qgestureevent.html), see [Gestures Overview](https://doc.qt.io/qt-5/gestures-overview.html#example-event-handling).

**Reply by Nerixyz** [reply]
Hmm, from my small testing, this only works on touchscreens but not on touchpads. It seems that QT doesn't really handle it or if any it's handled different for each platform.

**Reply by kornes** [reply]
qt6 adds new QNativeGestureEvent events which might help: Swipe and Pan, docs mention touchpad briefly https://doc.qt.io/qt-6.2/qnativegestureevent.html

**Reply by Nerixyz** [reply]
It's [already in Qt5](https://doc.qt.io/qt-5/qnativegestureevent.html), but I'm not sure if that would work. It seems to be mainly for macOS:

> The `QNativeGestureEvent` used to be only available on OS X; it's not really an API that's meant for wide consumption.

[Source](https://stackoverflow.com/a/40289162/16300717)

**Reply by kornes** [reply]
yea, saw some qt tracker comments mentioning lacking cross implementations too, 
but still worth to at least test QNativeGestureEvent, might aid your own win32 gestures handling and make macOS support easier

**Comment by Nerixyz** [comment]
I finally found out how these events are captured.

Both Firefox and Chrome use the [DirectManipulation API](https://learn.microsoft.com/en-us/windows/win32/directmanipulation/direct-manipulation-portal) on Windows.

I tried implementing it [on this branch](https://github.com/Nerixyz/chatterino2/tree/experiment/swipe), but I couldn't get it to play nicely with the scrollbar (either the scrollbar doesn't work _or_ DirectManipulation doesn't fire events anymore). If anyone knows how to fix this, feel free to reach out.

As a reference, the Firefox implementation can be found [here](https://searchfox.org/mozilla-central/source/widget/windows/DirectManipulationOwner.cpp).
